### PR TITLE
Add InfinityScans

### LINKED ISSUE #18681: Infinity scans
### Source name

Infinity scans 

### Source link

https://infinityscans.xyz/

### Source language

English 

### Other details

Could you please add the infinity scans. After flame scans went down this would be last website where we can read worthwhile manhuas

### Acknowledgements

- [X] I have checked that the extension does not already exist on the [website extensions list](https://tachiyomi.org/extensions/) or the app.
- [X] I have checked that the extension is not on [the removed sources list](https://github.com/tachiyomiorg/tachiyomi-extensions/blob/master/REMOVED_SOURCES.md).
- [X] I have searched the existing issues and this is a new ticket, **NOT** a duplicate or related to another open or closed issue.
- [X] I have checked that the extension does not already exist by searching the [GitHub repository](https://github.com/tachiyomiorg/tachiyomi-extensions/) and verified it does not appear in the code base.
- [X] I have written a meaningful title with the source name.
- [X] I will fill out all of the requested information in this form.

### PR DESCRIPTION
**Pull request by Secozzi**
Closes #18681

Checklist:

- [x] Updated `extVersionCode` value in `build.gradle` for individual extensions
- [ ] Updated `overrideVersionCode` or `baseVersionCode` as needed for all multisrc extensions
- [x] Referenced all related issues in the PR body (e.g. "Closes #xyz")
- [x] Added the `isNsfw = true` flag in `build.gradle` when appropriate
- [ ] Have not changed source names
- [ ] Have explicitly kept the `id` if a source's name or language were changed
- [x] Have tested the modifications by compiling and running the extension through Android Studio

### CODE REVIEW

File: src/en/infinityscans/src/eu/kanade/tachiyomi/extension/en/infinityscans/InfinityScans.kt [hunk 1]
```diff
@@ -0,0 +1,312 @@
...
+    override val lang = "en"
+
+    override val supportsLatest = true
+
+    override val client: OkHttpClient = network.cloudflareClient.newBuilder()
+        .addInterceptor { chain ->
+            val response = chain.proceed(chain.request())
+            if (response.code == 401) {
+                response.close()
+                throw Exception("Solve Captcha in WebView")
```
Comment (AwkwardPeak7) [inline src/en/infinityscans/src/eu/kanade/tachiyomi/extension/en/infinityscans/InfinityScans.kt:45]: throw `IOException` in an interceptor

File: src/en/infinityscans/src/eu/kanade/tachiyomi/extension/en/infinityscans/InfinityScans.kt [hunk 2]
```diff
@@ -0,0 +1,312 @@
...
+        .build()
+
+    override fun headersBuilder() = super.headersBuilder()
+        .add("Referer", "$baseUrl/")
+
+    private val json: Json by injectLazy()
+
+    // Popular
+
+    override fun fetchPopularManga(page: Int): Observable<MangasPage> =
```
Comment (AwkwardPeak7) [inline src/en/infinityscans/src/eu/kanade/tachiyomi/extension/en/infinityscans/InfinityScans.kt:59]: avoid overriding the fetch methods, as they depend on rxjava which will eventually™  be removed.

File: src/en/infinityscans/src/eu/kanade/tachiyomi/extension/en/infinityscans/InfinityScans.kt [hunk 3]
```diff
@@ -0,0 +1,312 @@
...
+    override fun latestUpdatesParse(response: Response): MangasPage =
+        throw UnsupportedOperationException("Not used")
+
+    // Search
+
+    override fun fetchSearchManga(page: Int, query: String, filters: FilterList): Observable<MangasPage> {
+        return client.newCall(searchMangaRequest(page, query, filters))
+            .asObservableSuccess()
+            .map { response ->
+                searchMangaParse(response, page)
```
Comment (AwkwardPeak7) [inline src/en/infinityscans/src/eu/kanade/tachiyomi/extension/en/infinityscans/InfinityScans.kt:85]: page number is in the url query parameter so can take it from there inside `searchMangaParse`

File: src/en/infinityscans/src/eu/kanade/tachiyomi/extension/en/infinityscans/InfinityScans.kt [hunk 4]
```diff
@@ -0,0 +1,312 @@
...
+        selectFirst("div:has(>b:matches($name:))")?.ownText()
+
+    private fun Element.getLinks(name: String): String? =
+        select("div:has(>b:matches($name:)) a")
+            .joinToString(", ", transform = Element::text).trim()
+            .takeIf { it.isNotBlank() }
+
+    // Chapters
+
+    override fun chapterListParse(response: Response): List<SChapter> {
```
Comment (AwkwardPeak7) [inline src/en/infinityscans/src/eu/kanade/tachiyomi/extension/en/infinityscans/InfinityScans.kt:224]: why not override `chapterListRequest` to make the correct post request there and avoid an unnecessary call? As you don't seem to be using the passed response.

File: src/en/infinityscans/src/eu/kanade/tachiyomi/extension/en/infinityscans/InfinityScans.kt [hunk 5]
```diff
@@ -0,0 +1,312 @@
...
+        ).execute().parseAs<ResponseDto<ChapterDataDto>>()
+
+        return chapterListData.result.chapters.map {
+            it.toSChapter(slug)
+        }
+    }
+
+    // Pages
+
+    override fun pageListParse(response: Response): List<Page> {
```
Comment (AwkwardPeak7) [inline src/en/infinityscans/src/eu/kanade/tachiyomi/extension/en/infinityscans/InfinityScans.kt:254]: same here, make required request in `pageListRequest`

File: src/en/infinityscans/src/eu/kanade/tachiyomi/extension/en/infinityscans/InfinityScansDto.kt [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+    val id: Int,
+    val title: String,
+    val sequence: Int,
+    val date: String,
+) {
+    fun toSChapter(slug: String) = SChapter.create().apply {
+        name = title
+
+        // Things like prologues mess up the sequence number
+        chapter_number = title.substringAfter("hapter ").toFloatOrNull() ?: sequence.toFloat()
```
Comment (AwkwardPeak7) [inline src/en/infinityscans/src/eu/kanade/tachiyomi/extension/en/infinityscans/InfinityScansDto.kt:61]: the app will fill the chapter_number so no need to set it yourself.

File: src/en/infinityscans/src/eu/kanade/tachiyomi/extension/en/infinityscans/InfinityScans.kt (hunk 4 above)
Comment (Secozzi) [reply src/en/infinityscans/src/eu/kanade/tachiyomi/extension/en/infinityscans/InfinityScans.kt:224]: The call isn't unnecessary, the site adds a long, seemingly random string as a path segment on redirect, so the chapterListRequest is going to redirect to the correct URL, which is needed for referer.

File: src/en/infinityscans/src/eu/kanade/tachiyomi/extension/en/infinityscans/InfinityScansDto.kt (hunk 6 above)
Comment (Secozzi) [reply src/en/infinityscans/src/eu/kanade/tachiyomi/extension/en/infinityscans/InfinityScansDto.kt:61]: I'm not really sure what you mean, as removing that lines just causes every chapter to have -1 as chapter_number

Comment (AwkwardPeak7) [reply src/en/infinityscans/src/eu/kanade/tachiyomi/extension/en/infinityscans/InfinityScansDto.kt:61]: well that's odd, it is supposed to set it
https://github.com/tachiyomiorg/tachiyomi/blob/master/app/src/main/java/eu/kanade/domain/chapter/interactor/SyncChaptersWithSource.kt#L104-L105

Comment (Secozzi) [reply src/en/infinityscans/src/eu/kanade/tachiyomi/extension/en/infinityscans/InfinityScansDto.kt:61]: It's showing -1 for both stable 0.14.7 and preview r6309 so I'll just leave it in

**Pull request merged**